### ISSUE TITLE
Emit TestStepStarted event on @BeforeAll and @AfterAll hooks

### ISSUE THREAD

**Issue by HardNorth** [cucumber/common#1834]
**Is your feature request related to a problem? Please describe.**
It look like `@BeforeAll` and `@AfterAll` hooks do not fire any events, at least [I don't see it in the code explicitly](https://github.com/cucumber/cucumber-jvm/pull/1876/files#diff-1a0fb5d4f12d2de943eb4d191d6cf810baf35f1d1c8e4da56fec7fc075c59e40R71-R87).

This makes impossible to report them in third-party report systems. At least in a straightforward way, without hacking that with AspectJ.

**Describe the solution you'd like**
Fire `TestStepStarted` event as you do for all other hooks.

**Comment by mpkorstanje** [comment]
@aslakhellesoy I don't think Cucumber currently has a message type suitable for before all/after all hooks.

Using `TestStep{Started,Finished}` would be inappropriate because there is no `testCaseStartedId`.

**Comment by aurelien-reeves** [comment]
From the perspective of cucumber-ruby, as documented here: https://github.com/cucumber/cucumber-ruby/tree/main/features/docs/writing_support_code/hooks#beforeall-and-afterall, BeforeAll and AfterAll hooks are not part of the execution of the tests. There are suppose to setup and clean-up things that are not related to cucumber. 

As @mpkorstanje said, I am not sure there would be an appropriate event for those.

We could maybe add new ones dedicated to such hooks?

**Comment by aslakhellesoy** [comment]
I think we would need new message types in https://github.com/cucumber/common/tree/main/messages/jsonschema - maybe `BeforeAllHookStarted` and `BeforeAllHookFinished`.

**Comment by mpkorstanje** [comment]
That would mean adding four messages `{Before,After{AllHook{Started,Finished}`. For regular hooks we don't even make that distinction. There is a lack of consistency there that bothers me.

**Comment by mattwynne** [comment]
What kind of reporting do you want to be able to do @HardNorth?

**Comment by HardNorth** [comment]
@mattwynne I'm not really sure what you asking. I am a developer of [Report Portal](https://reportportal.io). It collects all test information (structure, hooks, logs, etc.) And analyse it with a taste of AI. I implemented numerous of test framework integrations for our project. Without these hooks reporting we won't be able to reflect them properly in our application, hence if such hook failed it won't be counted which cause statistics issues. And these hooks are also invisible for our analyser as for now

**Comment by aslakhellesoy** [comment]
> That would mean adding four messages `{Before,After}AllHook{Started,Finished}`. For regular hooks we don't even make that distinction. There is a lack of consistency there that bothers me.

Good point. Perhaps it would be better to simply have `HookStarted` and `HookFinished`. We could have a property indicating what kind of hook it is, e.g `type = “BeforeAll” | “BeforeFeature” | “BeforeScenario” | “BeforeStep” | ….`. If we go down this route we should also consider replacing the use of `TestStepStarted` and `TestStepFinished` for regular (scenario) hooks.

**Comment by mattwynne** [comment]
In my mind, the original `{Before,After}{Scenario,Step}` hooks are just special cases of `TestSteps` since if they fail they can fail a test case. However a `BeforeAll` hook isn't part of any specific test case, so it feels like a different beast.

I'm curious how you were thinking about reporting on it @HardNorth. Would a failure of a `BeforeAll` hook be counted as a failure of _all_ the tests in the run? Or do you count these sorts of hooks as separate to the tests themselves?

**Comment by mpkorstanje** [comment]
The equivalent of a special test step for before hooks would be a special test case for before all hooks.

And a failure of a hook would be equivalent to failing the test run (same as with failure of a test case, worst result wins).

**Comment by HardNorth** [comment]
@mattwynne We don't report all tests as failed usually, just a single Item which failed. But for those who want to be extremely accurate I leave a public empty method, which can be overridden.
But without any event we won't even report a failed test run. That will be confusing for our users.

**Comment by aurelien-reeves** [comment]
> Good point. Perhaps it would be better to simply have `HookStarted` and `HookFinished`. We could have a property indicating what kind of hook it is, e.g `type = “BeforeAll” | “BeforeFeature” | “BeforeScenario” | “BeforeStep” | ….`. 

+1

> If we go down this route we should also consider replacing the use of `TestStepStarted` and `TestStepFinished` for regular (scenario) hooks.

+1 keeping in mind to add a proper deprecation notice before totally replacing those

**Comment by aslakhellesoy** [comment]
@aurelien-reeves I am not suggesting we deprecate `TestStepStarted` and `TestStepFinished`.

I am suggesting we use those messages only for Gherkin steps, and not for hooks.

**Comment by aurelien-reeves** [comment]
> @aurelien-reeves I am not suggesting we deprecate `TestStepStarted` and `TestStepFinished`.
> 
> I am suggesting we use those messages only for Gherkin steps, and not for hooks.

Yes, I understood 👌

Still, not using `TestStep{Started|Finished}` anymore with hooks remains a breaking change.

**Comment by aslakhellesoy** [comment]
Yes it would be a breaking change, but there isn't anything to deprecate.
